include/yaclib/algo/when_all.hpp

```diff
@@ -1,123 +1,19 @@
-          std::move(promise).Set();
-        } else {
-          std::move(promise).Set(std::vector<T>{});
-        }
-        return {std::move(future), nullptr};
-      }
-    }
-    return {std::move(future), new util::Counter<AllCombinator>{std::move(promise), size}};
-  }
-
-  void CallInline(void* context) noexcept final {
-    if (BaseCore::GetState() != BaseCore::State::HasStop) {
-      Combine(std::move(static_cast<ResultCore<T>*>(context)->Get()));
-    }
-  }
-
-  void Combine(util::Result<T>&& result) noexcept(std::is_void_v<T> || std::is_nothrow_move_assignable_v<T>) {
-    if (Base::_done.load(std::memory_order_acquire)) {
-      return;
-    }
-    auto state = result.State();
-    if (state == util::ResultState::Value) {
-      if constexpr (!std::is_void_v<T>) {
-        const auto ticket = AllCombinatorBase<FutureValue>::_ticket.fetch_add(1, std::memory_order_acq_rel);
-        AllCombinatorBase<FutureValue>::_results[ticket] = std::move(result).Value();
-      }
-      return;
-    }
-    if (Base::_done.exchange(true, std::memory_order_acq_rel)) {
-      return;
-    }
-    if (state == util::ResultState::Error) {
-      std::move(_promise).Set(std::move(result).Error());
-    } else {
-      std::move(_promise).Set(std::move(result).Exception());
-    }
-  }
-
-  ~AllCombinator() override {
-    if (!Base::_done.load(std::memory_order_acquire)) {
-      if constexpr (!std::is_void_v<T>) {
-        std::move(_promise).Set(std::move(AllCombinatorBase<FutureValue>::_results));
-      } else {
-        std::move(_promise).Set();
-      }
-    }
-  }
-
- private:
-  explicit AllCombinator(Promise<FutureValue> promise, [[maybe_unused]] size_t size = 0)
-      : BaseCore{BaseCore::State::Empty}, _promise{std::move(promise)} {
-    if constexpr (!std::is_void_v<T> && !IsArray) {
-      AllCombinatorBase<FutureValue>::_results.resize(size);
-    }
-  }
-
-  Promise<FutureValue> _promise;
-};
-
-template <size_t N, typename T, typename... Fs>
-void WhenAllImpl(util::Ptr<AllCombinator<T, N>>& combinator, Future<T>&& head, Fs&&... tail) {
-  head.GetCore()->SetCallbackInline(combinator);
-  std::move(head).Detach();
-  if constexpr (sizeof...(tail) != 0) {
-    WhenAllImpl(combinator, std::forward<Fs>(tail)...);
-  }
-}
-
-}  // namespace detail
 
 /**
  * Create \ref Future which will be ready when all futures are ready
  *
  * \tparam It type of passed iterator
  * \tparam T type of all passed futures
  * \param begin, size the range of futures to combine
- * \return Future<vector<T>>
+ * \return Future<std::vector<T>>
  */
-template <typename It, typename T = util::detail::FutureValueT<typename std::iterator_traits<It>::value_type>>
+template <WhenPolicy P = WhenPolicy::FirstFail, typename It,
+          typename T = util::detail::FutureValueT<typename std::iterator_traits<It>::value_type>>
 auto WhenAll(It begin, size_t size) {
   auto [future, combinator] = detail::AllCombinator<T>::Make(size);
```

Comment:
Add for all WhenAll static_assert(p == WhenPolicy::FirstFail, "TODO(Ri7ay) Add other policy for WhenAll")